### ISSUE TITLE
Allow `_` in `match` to appear earlier on

### ISSUE THREAD

**Issue by brson** [lalrpop/lalrpop#325]
I have a grammar  for which the match statement looks like

```
match {
    "access_log",
    "client_body_temp_path",
    "default_server",
    "error_log",
    "events",
    "fastcgi_temp_path",
    "http",
    "index",
    "listen",
    "pid",
    "proxy_temp_path",
    "root",
    "scgi_temp_path",
    "server",
    "server_name",
    "user",
    "uwsgi_temp_path",
    "worker_connections",
    "worker_processes",
    "worker_rlimit_nofile",
    ";",
    "{",
    "}",
} else {
    r"[[:digit:]]+" => NUMBER,
} else {
    r"[[:alnum:]_]+" => NAME,
} else {
    r"[[:alnum:]/.]+" => PATH,
}
```

I sure wish I didn't have to write all those simple literals out, but I can't figure out how to avoid them. I can't put the wildcard `_` in the first arm of `match`

```
match {
    _
} else {
    r"[[:digit:]]+" => NUMBER,
} else {
    r"[[:alnum:]_]+" => NAME,
} else {
    r"[[:alnum:]/.]+" => PATH,
}
```

That doesn't work:

```
brian@DESKTOP-UCV672I:~/engine-arr⟫ cargo run -- -c examples/blocks.conf -T
   Compiling arr v0.1.0 (file:///home/brian/engine-arr)
error: failed to run custom build command for `arr v0.1.0 (file:///home/brian/engine-arr)`
process didn't exit successfully: `/home/brian/engine-arr/target/debug/build/arr-f72285b38b3e56b8/build-script-build` (exit code: 1)
--- stdout
processing file `/home/brian/engine-arr/src/config_parser.lalrpop`
/home/brian/engine-arr/src/config_parser.lalrpop:11:5: 11:5 error: Catch all must be final item

--- stderr
      _
      ^

```

The "catch all" must be the final item. So this can be written:

```
match {
    r"[[:digit:]]+" => NUMBER,
} else {
    r"[[:alnum:]_]+" => NAME,
} else {
    r"[[:alnum:]/.]+" => PATH,
} else {
    _
}
```

It doesn't do what I want it to do though. My simple tokens in the grammar seemingly get lexed as regexes, resulting in some confusing error messages:

```
   Compiling arr v0.1.0 (file:///home/brian/engine-arr)
    Finished dev [unoptimized + debuginfo] target(s) in 8.75 secs
     Running `target/debug/arr -c examples/blocks.conf -T`
err: Unrecognized token `pid` found at 0:3
Expected one of "error_log", "events", "http", "pid", "user", "worker_processes" or "worker_rlimit_nofile"
```

So at  the moment I'm listing every terminal in my language in my `match` statement.

Not sure what I'm asking for, but I wish I wasn't as confused. I wish I didn't have to list all my tokens.

**Comment by nikomatsakis** [comment]
@brson

>  Not sure what I'm asking for, but I wish I wasn't as confused. I wish I didn't have to list all my tokens.

D'oh. I hadn't noticed that the `_` wasn't actually working for you. Hmm, it sounds like you want the catch all to take top priority. I'm not sure why we required it to be the final item, actually. I can't remember if there is a good reason for this.

**Comment by nikomatsakis** [comment]
Some investigation:

The AST that we construct for match items is here:

https://github.com/lalrpop/lalrpop/blob/896ab62e039a3a7f5f986bed102864c8b813f740/lalrpop/src/grammar/parse_tree.rs#L85-L92

This gets processed here into one of these data structures, where the catch all becomes just a boolean:

https://github.com/lalrpop/lalrpop/blob/896ab62e039a3a7f5f986bed102864c8b813f740/lalrpop/src/normalize/token_check/mod.rs#L88-L111

I *think* that we should change `match_catch_all` from a boolean to `Option<usize>` (or maybe `Option<Precedence>`, and then record the value of `precedence` at this point:

https://github.com/lalrpop/lalrpop/blob/896ab62e039a3a7f5f986bed102864c8b813f740/lalrpop/src/normalize/token_check/mod.rs#L132-L134

Then here, where we process new terminals:

https://github.com/lalrpop/lalrpop/blob/896ab62e039a3a7f5f986bed102864c8b813f740/lalrpop/src/normalize/token_check/mod.rs#L180-L188

we would change the precedence from `sym.base_precedence()`:

https://github.com/lalrpop/lalrpop/blob/896ab62e039a3a7f5f986bed102864c8b813f740/lalrpop/src/normalize/token_check/mod.rs#L185

to match this equation:

https://github.com/lalrpop/lalrpop/blob/896ab62e039a3a7f5f986bed102864c8b813f740/lalrpop/src/normalize/token_check/mod.rs#L160

(We might want to extract a helper.)

The basic idea is that `MatchEntry` precedence levels are an integer. For each block in the `match { } else { }` sequence, there are two levels -- one for fixed strings and one for regex. So the `usize` we recorded is the block, and hence we have to (a) multiply by 2 and then (b) add the offset of 1 for fixed strings (that is the `base_precedence` returned by symbol).

**Comment by k0pernicus** [comment]
I can work on it tonight if no one's working on it.

**Comment by nikomatsakis** [comment]
@k0pernicus great! ping me if you run into any questions, and/or reach out on gitter

**Comment by k0pernicus** [comment]
@nikomatsakis Thanks. I will check the code of conduct and the documentation to know how to build, test and run the project. Next, I will start working on this issue.

**Comment by nikomatsakis** [comment]
@k0pernicus I hope the info you want is there :) let me know otherwise

**Comment by k0pernicus** [comment]
@nikomatsakis I made (quickly) the tutorial, and I know understand what is the goal of this issue.
I modified the `catch_all` attribute from `bool` to `Option<Precedence>`, and set it to `None` if the user did not provides any data out of the `match` block (otherwise, I set it to `Some(Precedence(precedence))`.
But, I am a little confused with your change in `add_literal_from_grammar` ( more precisely, the new computation of `precedence` line 185, in order to match the computation of the same attribute line 160 ). Indeed, I don't understand how to get the current number of precedence (`match_group_precedence` if I am correct ?) in order to compute the new one in `add_literal_from_grammar`...
Is `match_group_precedence`, in this case, is defining the last precedence (max) ?

Thanks a lot for your help :-)

**Comment by k0pernicus** [comment]
By the way, I created a new test in `lalrpop-test`, which is a modified version of an example in the tutorial:
```
grammar;

match {
    _
} else {
    r"\w+"
}

pub Term = {
    Num,
    "(" <Term> ")",
    "22" => format!("Twenty-two!"),
    r"\w+" => format!("Id({})", <>), // <-- we added this
};

Num: String = r"[0-9]+" => <>.to_string();
```
Using this test (just a modification in the `match` block, considering the example given in the tutorial for predecence), we can know if precedence has been correctly setted if [0..21] & [23.. +inf] returns the correct integer, "22" returns "Twenty-two!", and another alphanumeric match returns "Id(...)".  
As I am new in the project, I can't say if this two-level test is not covering enough this feature, or if I can use already written unit tests instead of writing a new one...

Otherwise, I think I need to modify other parts of the code like this:
https://github.com/lalrpop/lalrpop/blob/896ab62e039a3a7f5f986bed102864c8b813f740/lalrpop/src/normalize/prevalidate/mod.rs#L78-L90

In this code, we check if catch all is the final item of the final block - which is not correct anymore.
Do I need to modify or to remove this part of code?

**Comment by nikomatsakis** [comment]
@k0pernicus 

> ndeed, I don't understand how to get the current number of precedence (match_group_precedence if I am correct ?) in order to compute the new one in add_literal_from_grammar...
Is match_group_precedence, in this case, defines the precedence of the last entry/literal ?

You want to get the precedence of the "catch-all" -- you should be able to change this function:

https://github.com/lalrpop/lalrpop/blob/896ab62e039a3a7f5f986bed102864c8b813f740/lalrpop/src/normalize/token_check/mod.rs#L167-L194

to something like this:

```rust
    fn add_literal_from_grammar(&mut self, sym: TerminalLiteral, span: Span) -> NormResult<()> {
        // Already saw this literal, maybe in a match entry, maybe in the grammar.
        if self.match_user_names
               .contains(&TerminalString::Literal(sym.clone())) {
            return Ok(());
        }

        // XXX this is changed
        let precedence = match self.catch_all {
            Some(p) => p,
            None => {
                return_err!(span,
                        "terminal `{}` does not have a match mapping defined for it",
                        sym);
            }
       };

        self.match_user_names
            .insert(TerminalString::Literal(sym.clone()));

        self.match_entries
            .push(MatchEntry {
                      precedence: precedence * 2 + sym.base_precedence(), // XXX this too
                      match_literal: sym.clone(),
                      user_name: TerminalString::Literal(sym.clone()),
                  });

        self.spans.insert(sym, span);

        Ok(())
}
```

**Comment by nikomatsakis** [comment]
>  As I am new in the project, I can't say if this two-level test is not covering enough this feature, or if I can use already written unit tests instead of writing a new one...

This test looks good. It'd be good to put a comment in the text explaining what it is trying to test and how. There may be some more targeted unit tests that could work, but an end-to-end test is always good.

**Comment by nikomatsakis** [comment]
Note that you will want to run rustfmt too, thanks to https://github.com/lalrpop/lalrpop/pull/327

**Comment by k0pernicus** [comment]
Thanks @nikomatsakis - I just changed my solution with your code proposition (in order to get the `precedence` value).

So, are you agree that tests `match_catch_all_first_of_last` and `match_catch_all_last_of_first` must be OK now, as we accept the wildcard in the first arm of `match`?
Or, maybe we can replace the error message `Catch all must be final item` by something like `"X" is already setted in the wildcard symbol "_"`?

**Comment by k0pernicus** [comment]
Those tests are defined here:
https://github.com/lalrpop/lalrpop/blob/ebc1e5049afd41ffeda2f9cf63fca229e88c782c/lalrpop/src/normalize/prevalidate/test.rs#L109-L116

and here:
https://github.com/lalrpop/lalrpop/blob/ebc1e5049afd41ffeda2f9cf63fca229e88c782c/lalrpop/src/normalize/prevalidate/test.rs#L118-L125

**Comment by brson** [comment]
Woohoo. Thanks for working on this @nikomatsakis and @k0pernicus.

**Comment by nikomatsakis** [comment]
@k0pernicus d'oh, missed your updates (please do ping on gitter, I get overwhelmed easily =)

Regarding those tests -- it seems ok to support the catch all in any place, yes, as long as it appears at most once. I don't think it's location matters, right (within a match arm, I mean)? Wherever you put it, the entries from the grammar will be added as regex or literal priority, and all things are equal within a match arm except for that.

**Comment by Henriquelay** [comment]
Is this still going on? I tried to have `_` match before a lower precedence term, and this still happens.
I'm not skilled enough to update the PR after these years.

**Comment by k0pernicus** [comment]
Hi @Henriquelay,
This is my entire fault, I did not checked the latest comments on PR + ping Niko to review the latest changes...

I will check the following days if this PR is still valid (and if another person did not implement another solution for this...), and let you know in the next days.

**Comment by Pat-Lafon** [comment]
I believe this has finally landed in #807 so I will be closing this issue. If that doesn't fully address this use case then we can reopen. Thanks!
